src/H5private.h

```diff
@@ -1295,19 +1295,34 @@ H5_DLL herr_t H5CX_pop(bool update_dxpl_props);
             FUNC_ENTER_API_PUSH(err);                                                                        \
             {
 
+/*
+ * Use this macro for "developer" API functions that re-enter the library from
+ *      a plugin / connector / filter / etc, and shouldn't (need to) perform
+ *      initialization of the library or an interface, just perform tracing, etc.
+ *      Examples are: H5allocate_memory, public VOL callback
+ *      wrappers (e.g. H5VLfile_create, H5VLdataset_read, etc.), public VFL
+ *      callback wrappers (e.g. H5FDopen, H5FDwrite, etc.), etc.
+ *
+ */
+#define FUNC_ENTER_API_REENTER                                                                               \
+    {                                                                                                        \
+        {                                                                                                    \
+            {                                                                                                \
+                FUNC_ENTER_API_COMMON                                                                        \
+                H5_PUSH_FUNC                                                                                 \
+                {
+
 /*
  * Use this macro for API functions that shouldn't perform _any_ initialization
  *      of the library or an interface, just perform tracing, etc.  Examples
- *      are: H5allocate_memory, H5is_library_threadsafe, public VOL callback
- *      wrappers (e.g. H5VLfile_create, H5VLdataset_read, etc.), etc.
+ *      are: H5is_library_threadsafe, H5VLretrieve_lib_state, etc.
  *
  */
 #define FUNC_ENTER_API_NOINIT                                                                                \
     {                                                                                                        \
         {                                                                                                    \
             {                                                                                                \
                 FUNC_ENTER_API_COMMON                                                                        \
-                H5_PUSH_FUNC                                                                                 \
```

Comment:
What's the motivation for removing utility functions that still use `_NOINIT` from the function stack?